File: 2017/qualification/bathroom/Bathroom.java
import java.util.PriorityQueue;
import java.util.Collections;
import java.util.Scanner;
public class Bathroom {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        int times = Integer.parseInt(sc.nextLine());

        for (int i = 1; i <= times; i++) {
            String[] tupleLine = sc.nextLine().split(" ");

            int stalls = Integer.parseInt(tupleLine[0]);
            int people = Integer.parseInt(tupleLine[1]);

            int max = stalls;
            int min = stalls;
            
            PriorityQueue<Integer> bath = new PriorityQueue<Integer>(1, Collections.reverseOrder());
            
            bath.offer(stalls);

            for (int j = 1; j <= people; j++) {
                int temp = bath.poll();
                max = temp / 2;
                min = temp % 2 == 0 ? temp / 2 - 1 : temp / 2;

             bath.offer(max);
             bath.offer(min);
            }

            System.out.println("Case #" + i + ": " + max + " " + min);
        }
    }
}

File: 2017/qualification/bathroom/bathroom.py
import queue

def updateMaxMin(max, min, bath):
    return 1

def solver (stalls, people):

    max = stalls
    min = stalls

    # full
    if (stalls == people):
        return [0,0]

    # start bath
    bath = queue.PriorityQueue()
    bath.put((-stalls, stalls))

    for i in range(people):
        temp = bath.get()[1]
        max = int(temp / 2)
        if temp % 2 == 0:
            min = int(temp / 2 - 1)
        else:
            max = int(temp / 2)

        bath.put((-max,max))
        bath.put((-min,min))
    
    return [max,min]


times = int(input())
for i in range(1, times + 1):
    stalls, people = (int(j) for j in input().split(' '))

    answer = solver (stalls, people) 
    print("Case #{}: {} {}".format(str(i), str(answer[0]), str(answer[1])))
